### PR TITLE
Vapor 4

### PR DESCRIPTION
**Pull request by Maxim-Inv**
Converted to support last Vapor 4 beta

### CODE REVIEW

**Comment by 0xTim** [comment]
Can you point this to the `4` to start with? We won't merge it in to master until Vapor 4 is tagged

File: Package.swift [hunk 1]
```diff
@@ -1,18 +1,27 @@
-// swift-tools-version:4.0
+// swift-tools-version:5.1
 
 import PackageDescription
 
 let package = Package(
     name: "Imperial",
+    platforms: [
+       .macOS(.v10_14)
```
Comment (0xTim) [inline Package.swift]: This will need to be 10.15 if we're relying on JWT

File: Sources/Imperial/Helpers/Request+Imperial.swift [hunk 2]
```diff
@@ -10,29 +10,29 @@ extension Request {
...
+//        let token = try service.tokenPrefix + req
+//            .accessToken()
+//
+//        return self.make(Client.self).get(URI(string: url), headers: ["Authorization": token]).flatMap { response in
+//            return try! model.create(from: response).map { instance in
+//                try session.set("imperial-\(model)", to: instance)
+//                return instance
+//            }
+//        }
+        fatalError()
```
Comment (0xTim) [inline Sources/Imperial/Helpers/Request+Imperial.swift]: This needs to be implemented I'm guessing?

**Comment by Maxim-Inv** [comment]
Sorry, I mixed up.

**Comment by 0xTim** [comment]
No worries, you can ignore my last comment then since that's not something you've done. Feel free to implement though!

File: Package.swift (hunk 1 above)
Comment (Maxim-Inv) [reply Package.swift]: done

File: Sources/Imperial/Helpers/Request+Imperial.swift (hunk 2 above)
Comment (Maxim-Inv) [reply Sources/Imperial/Helpers/Request+Imperial.swift]: done

File: Sources/Imperial/Routing/FederatedServiceRouter.swift [hunk 3]
```diff
@@ -58,23 +58,16 @@ public protocol FederatedServiceRouter {
...
-//            guard let authenticateCallback = authenticateCallback else {
-//                return req.eventLoop.makeSucceededFuture(redirect)
-//            }
-//            return try authenticateCallback(req).map(to: Response.self) { _ in
-//                return redirect
-//            }
-//        }
-        fatalError()
+    public func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, on router: RoutesBuilder) throws {
+		router.get(callbackURL.pathComponents, use: callback)
```
Comment (0xTim) [inline Sources/Imperial/Routing/FederatedServiceRouter.swift]: Can you add the (commented out) code that sets the `callbackPath` to HTTPS if needed?

File: Sources/Imperial/Services/Facebook/FacebookRouter.swift [hunk 4]
```diff
@@ -23,38 +23,36 @@ public class FacebookRouter: FederatedServiceRouter {
...
+			return try response.content.get(String.self, at: ["access_token"])
+		}
     }
     
     public func callback(_ request: Request) throws -> EventLoopFuture<Response> {
         return try self.fetchToken(from: request).flatMap { accessToken in
             let session = request.session
             do {
-                try session.set("access_token", to: accessToken)
+				try session.setAccessToken(accessToken)
```
Comment (0xTim) [inline Sources/Imperial/Services/Facebook/FacebookRouter.swift]: Weird spacing issues here?

File: Sources/Imperial/Helpers/Sessions+Imperial.swift [hunk 5]
```diff
@@ -35,7 +35,7 @@ extension Session {
     ///
     /// - Parameter token: the access token to store on the session
     public func setAccessToken(_ token: String) throws {
-        try set(Keys.token, to: token)
+		try set(Keys.token, to: token)
```
Comment (0xTim) [inline Sources/Imperial/Helpers/Sessions+Imperial.swift]: Not sure why this is so indented 🤔

Comment (Maxim-Inv) [reply Sources/Imperial/Helpers/Sessions+Imperial.swift]: I'm using tab indentation. Fixed.

File: Sources/Imperial/Services/Facebook/FacebookRouter.swift (hunk 4 above)
Comment (Maxim-Inv) [reply Sources/Imperial/Services/Facebook/FacebookRouter.swift]: Fixed.

File: Sources/Imperial/Routing/FederatedServiceRouter.swift (hunk 3 above)
Comment (Maxim-Inv) [reply Sources/Imperial/Routing/FederatedServiceRouter.swift]: These routes depend from the vapor configuration. Is it possible to make certains path to HTTPS if the rest of them are HTTP?

**Pull request merged**